ml/train/losses/losses.go

```diff
@@ -398,3 +398,211 @@ func MakeHuberLoss(delta float64) LossFn {
 		return loss
 	}
 }
+
+// Computes the pairwise distance matrix with numerical stability.
+
+//     output[i, j] = || embeddings[i, :] - embeddings[j, :] ||_2
+
+//     Args:
+//       embeddings: 2-D Tensor of size `[number of data, embeddings dimension]`.
+//       squared: Boolean, whether or not to square the pairwise distances.
+
+// Returns:
+//
+//	pairwise distance: 2-D Tensor of size `[number of data, number of data]`.
+func pairwiseDistance(embeddings *Node, squared bool) *Node {
+	// Distances ||a - b||^2 = ||a||^2  - 2 <a, b> + ||b||^2
+	pairwiseDistancesSquared := Add(
+		Add(
+			L2NormSquare(embeddings, 1),
+			L2NormSquare(Transpose(embeddings, 0, 1), 0),
+		),
+		MulScalar(MatMul(embeddings, Transpose(embeddings, 0, 1)), -2.0),
+	)
+	// Deal with numerical inaccuracies. Set small negatives to zero.
+	pairwiseDistancesSquaredZeros := Max(pairwiseDistancesSquared, ZerosLike(pairwiseDistancesSquared))
+	// Optionally take the sqrt.
+	pairwiseDistances := pairwiseDistancesSquaredZeros
+	if !squared {
+		// Get the mask where the zero distances are at.
+		zeros := ZerosLike(pairwiseDistancesSquaredZeros)
+		errorMask := LessOrEqual(pairwiseDistancesSquaredZeros, zeros)
+		epsilon := epsilonForDType(pairwiseDistancesSquaredZeros.Graph(), pairwiseDistancesSquaredZeros.DType())
+		pairwiseDistancesSqrt := Sqrt(Where(errorMask, epsilon, pairwiseDistancesSquaredZeros))
+		//  Undo conditionally adding epsilon
+		pairwiseDistances = Where(errorMask, zeros, pairwiseDistancesSqrt)
+	}
+
+	//  Explicitly set diagonals to zero.
+	return Where(
+		Diagonal(pairwiseDistances.Graph(), pairwiseDistances.Shape().Dim(0)),
+		ZerosLike(pairwiseDistances),
+		pairwiseDistances,
+	)
+}
+
+// Computes the angular distance matrix.
+
+//     output[i, j] = 1 - cosine_similarity(feature[i, :], feature[j, :])
+
+//     Args:
+//       feature: 2-D Tensor of size `[number of data, feature dimension]`.
+
+// Returns:
+//
+//	angular_distances: 2-D Tensor of size `[number of data, number of data]`.
+func angularDistance(embeddings *Node) *Node {
+	// normalize input
+	embeddingsN := L2Normalize(embeddings, 1)
+	// create adjacent matrix of cosine similarity
+	angularDistances := OneMinus(MatMul(embeddingsN, Transpose(embeddingsN, 0, 1)))
+	// ensure all distances > 1e-16
+	angularDistancesZeros := MaxScalar(angularDistances, 0.0)
+	return angularDistancesZeros
+}
+
+// Computes the axis wise maximum over chosen elements.
+
+//     Args:
+//       data: 2-D float `Tensor` of shape `[n, m]`.
+//       mask: 2-D Boolean `Tensor` of shape `[n, m]`.
+//       dim: The dimension over which to compute the maximum.
+
+// Returns:
+//
+//	masked_maximums: N-D `Tensor`.
+//	  The maximized dimension is of size 1 after the operation.
+func maskedMaximum(x, mask *Node, dim int) *Node {
+	axisMinimums := ReduceAndKeep(x, ReduceMin, dim)
+	return Add(
+		ReduceAndKeep(
+			Mul(
+				Sub(x, axisMinimums),
+				ConvertDType(mask, x.DType())),
```

Comment:
This is a tricky one.

The trick of multiplying by a mask converted to a float can lead to NaNs: if the masked value is NaN or Inf, when multiplied by zero, it is still NaN and Inf. I've seen this causing lots of headache during trainining, where things masked out were causing NaNs/Infs 😄 

If there is no way around, its ok ... but maybe there is ?

Btw, there is a MaskedReduceMax, not sure it would work here -- I have to leave but I'll review in more details later.